inference/core/workflows/core_steps/sampling/identify_outliers/v1.py

```diff
@@ -0,0 +1,197 @@
+        R = np.linalg.norm(sum_vec)
+        if R == 0:
+            # All embeddings canceled out, no direction
+            return None, None
+
+        mu = sum_vec / R
+        r_bar = R / n
+
+        # Approximate kappa:
+        # For d>2, a known approximation:
+        # kappa ≈ r_bar*(d - r_bar^2)/(1 - r_bar^2)
+        d = float(d)
+        if r_bar < 1.0:
+            kappa = (r_bar * (d - r_bar**2)) / (1 - r_bar**2)
+        else:
+            # Degenerate case: all points are identical
+            kappa = np.inf
+
+        return mu, kappa
+
+    def run(
+        self,
+        embedding: List[float],
+        threshold_percentile: float,
+        warmup: int,
+        window_size: int,
+    ) -> BlockResult:
+        # Convert to np array
+        embedding = np.array(embedding, dtype=float)
+        # Normalize embedding for vMF
+        norm = np.linalg.norm(embedding)
+        if norm == 0:
+            # If zero vector, skip normalization
+            embedding_normed = embedding
+        else:
+            embedding_normed = embedding / norm
+
+        self.samples += 1
+        warming_up = self.samples < warmup
+
+        # Store normalized embedding for vMF:
+        self.all_embeddings.append(embedding_normed)
+        if len(self.all_embeddings) > window_size:
+            # Remove oldest embedding
+            self.all_embeddings.pop(0)
+
+        # If we're still in warmup, we cannot decide outliers yet
+        if warming_up:
+            return {
+                "is_outlier": False,
+                "percentile": 0.5,
+                "warming_up": True,
+            }
+
+        # Fit vMF parameters based on all embeddings so far
+        all_emb_array = np.array(self.all_embeddings)
+        mu, kappa = self._fit_vmf_parameters(all_emb_array)
+        if mu is None or kappa is None:
+            # Fallback if we cannot fit (e.g. all embeddings identical)
+            mu = embedding_normed
+            kappa = 0.0
+
+        # Compute alignment score with the current mean direction
+        t_new = np.dot(mu, embedding_normed)
+
+        # Compute empirical percentile of t_new relative to historical t_i
+        # Using all previous embeddings (excluding the current one if desired)
+        t_values = np.einsum("ij,j->i", all_emb_array, mu)
+        # Sort t-values to find percentile
+        sorted_t = np.sort(t_values)
+        rank = np.searchsorted(sorted_t, t_new, side="left")
+        percentile = rank / len(sorted_t)
+
+        # Determine outlier based on percentile thresholds
+        is_outlier = (percentile < threshold_percentile) or (
+            percentile > (1 - threshold_percentile)
+        )
+
+        # print(is_outlier, f"{round(percentile*100)}%", warming_up)
+
+        # We retain fields "average" and "std" for compatibility, though they are less meaningful
+        # now. We can still return the SMA/EMA-based averages/std for debugging, or just zeros.
+        return {
```

Comment:
comment does match whats actually returned?

Reply:
Oops no, I changed the output format when I removed those. Forgot the comment. Will update